rest/route/agent.go

```diff
@@ -1750,3 +1751,94 @@ func (h *awsAssumeRole) Run(ctx context.Context) gimlet.Responder {
 		Expiration:      creds.Expiration.String(),
 	})
 }
+
+// POST /rest/v2/task/{task_id}/aws/s3
+// This route is used to retrieve credentials for an S3 bucket for a task.
+// s3.put and s3.get call this route when the command is targetting an
+// internal bucket.
+type awsS3 struct {
```

Comment:
Do we actually need a different route for this also? This seems the same as the assume_role route, but with an additional step if callerARN isn't provided, and the bucket. Should we consider just adding some boolean logic to the existing route?

Reply:
This route acts a bit different. `callerARN` isn't user-provided and is from Evergreen's IRSA role. This route is creating credentials that have restricted permissions to s3 but it AssumeRole Evergreen's IRSA role.

Reply:
I see that it behaves differently **up until ** we make the assume role call -- then it just does the same thing that assume role does. So couldn't we add a section to the assume_role route that has some boolean we pass in, like "generate_for_task" that creates the caller ARN and policy (the way that you do below) instead of using a passed in version. (We can take this to DM, but I'm not totally seeing why that isn't reasonable, since these do both assume role at the end of the day, and return the same values)

Alternatively if that's really not a good idea, having some helper function that's like getAPIAssumeRole that both the assume_role and s3_credentials route could call that takes in the AssumeRoleOptions as a parameter might be nice, so that part isn't duplicative at least.

Reply:
I see that it behaves differently **up until ** we make the assume role call -- then it just does the same thing that assume role does. So couldn't we add a section to the assume_role route that has some boolean we pass in, like "generate_for_task" that creates the caller ARN and policy (the way that you do below) instead of using a passed in version. (We can take this to DM, but I'm not totally seeing why that isn't reasonable, since these do both assume role at the end of the day, and return the same values)

Alternatively if that's really not a good idea, having some helper function that's like getAPIAssumeRole that both the assume_role and s3_credentials route could call that takes in the AssumeRoleOptions as a parameter might be nice, so that part isn't duplicative at least.

Reply:
I see that it behaves differently **up until ** we make the assume role call -- then it just does the same thing that assume role does. So couldn't we add a section to the assume_role route that has some boolean we pass in, like "generate_for_task" that creates the caller ARN and policy (the way that you do below) instead of using a passed in version. (We can take this to DM, but I'm not totally seeing why that isn't reasonable, since these do both assume role at the end of the day, and return the same values)

Alternatively if that's really not a good idea, having some helper function that's like getAPIAssumeRole that both the assume_role and s3_credentials route could call that takes in the AssumeRoleOptions as a parameter might be nice, so that part isn't duplicative at least.